File: learnLeetCode/Day35/Part5/palindrome_min_cuts.py
def min_cut(s: str) -> int:
    n = len(s)
    cut = [0]*(n+1)
    for i in range(n+1):
        cut[i] = i-1
    for c in range(n):
        # odd length
        l=r=c
        while l>=0 and r<n and s[l]==s[r]:
            cut[r+1] = min(cut[r+1], 1 + cut[l])
            l-=1; r+=1
        # even length
        l=c; r=c+1
        while l>=0 and r<n and s[l]==s[r]:
            cut[r+1] = min(cut[r+1], 1 + cut[l])
            l-=1; r+=1
    return cut[n]

if __name__ == "__main__":
    print(min_cut("aab"))  # Expected 1 (aa|b)

File: learnLeetCode_remote/Day35/Part5/palindrome_min_cuts.cpp
#include <bits/stdc++.h>
using namespace std;

int min_cut(const string& s){
    int n = (int)s.size();
    vector<int> cut(n+1);
    for(int i=0;i<=n;++i) cut[i]=i-1; // cut[0]=-1
    for(int c=0;c<n;++c){
        int l=c,r=c; // odd
        while(l>=0 && r<n && s[l]==s[r]){
            cut[r+1] = min(cut[r+1], 1 + cut[l]);
            --l; ++r;
        }
        l=c; r=c+1; // even
        while(l>=0 && r<n && s[l]==s[r]){
            cut[r+1] = min(cut[r+1], 1 + cut[l]);
            --l; ++r;
        }
    }
    return cut[n];
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout << min_cut(string("aab")) << "\n"; // Expected 1
    return 0;
}
